Source/sdc/Castro_sdc_util.H

```diff
@@ -0,0 +1,882 @@
+               GpuArray<Real, NUM_STATE>& U_new,
+               GpuArray<Real, NUM_STATE> const& C,
+               const int sdc_iteration) {
+
+    // The purpose of this function is to solve the system the
+    // approximate system dU/dt = R + C using the VODE ODE
+    // integrator.
+    // The solution we get here will then be used as the 
+    // initial guess to the Newton solve on the real system.
+
+    // We will do the implicit update of only the terms that 
+    // have reactive sources
+    //
+    // 0 : rho 
+    // 1:NumSpec : species
+    // NumSpec+1 : (rho E) or (rho e)
+
+#if (INTEGRATOR == 0)
+
+    // The tolerance we are solving to may depend on the
+    // iteration
+    auto relax_fac = std::pow(sdc_solver_relax_factor, sdc_order - sdc_iteration - 1);
+    auto tol_dens = sdc_solver_tol_dens * relax_fac;
+    auto tol_spec = sdc_solver_tol_spec * relax_fac;
+    auto tol_ener = sdc_solver_tol_ener * relax_fac;
+
+    // Update the momenta for this zone -- they don't react
+    for (int n = 0; n < AMREX_SPACEDIM; ++n) {
+        U_new[UMX+n] = U_old[UMX+n] + dt_m * C[UMX+n];
+    }
+
+    // Now only save the subset that participates in the
+    // nonlinear solve -- note: we include the old state
+    // in f_source
+
+    // load rpar
+
+    // If we are solving the system as an ODE, then we are 
+    // solving
+    //   dU/dt = R(U) + C
+    // so we simply pass in C
+    GpuArray<Real, NumSpec+2> C_react;
+
+    C_react[0] = C[URHO];
+    for (int n = 0; n < NumSpec; ++n) {
+        C_react[1+n] = C[UFS+n];
+    }
+    C_react[NumSpec+1] = C[UEINT];
+
+    dvode_t dvode_state;
+    burn_t burn_state;
+
+    burn_state.rho = C_react[0];
+    for (int n = 0; n < NumSpec; ++n) {
+        burn_state.xn[n] = C_react[1+n];
+    }
+
+    // set the momenta to be U_new
+    GpuArray<Real, AMREX_SPACEDIM> mom_source;
+    for (int n = 0; n < AMREX_SPACEDIM; ++n) {
+        mom_source[n] = U_new[UMX+n];
+    }
+
+#if defined(SDC_EVOLVE_ENERGY)
+    burn_state.umx_orig = U_new[UMX];
+    burn_state.umy_orig = U_new[UMY];
+    burn_state.umz_orig = U_new[UMZ];
+#endif
+
+    // temperature will be used as an initial guess in the EOS
+    burn_state.T = U_old[UTEMP];
+
+    // we are always solving for rhoe with the VODE predict
+    burn_state.e = U_new[UEDEN];
+
+    // store the subset for the nonlinear solve. We only 
+    // consider (rho e), not (rho E). This is because at 
+    // present we do not have a method of updating the 
+    // velocities during the multistep integration.
+
+    // Also note that the dvode_state is 1-based, but we'll
+    // access it as 0-based in our implementation of the 
+    // RHS routine
+
+    dvode_state.y(1) = U_old[URHO];
+    for (int n = 0; n < NumSpec; ++n) {
+        dvode_state.y(2+n) = U_old[UFS+n];
+    }
+    dvode_state.y(NumSpec+2) = U_old[UEINT];
+
+    // // set the maximum number of steps allowed
+    // dvode_state.MXSTEP = 25000;
+
+    dvode_state.t = 0.0_rt;
+
```

Comment:
We also need to set `dvode_state.HMXI` here.  You can either set it to 0 or to `1.0_rt / ode_max_dt`